mountpoint-s3/src/fs.rs

```diff
@@ -58,91 +58,84 @@ where
     inode: Inode,
     full_key: String,
     object_size: u64,
-    typ: FileHandleType<Client, Prefetcher>,
+    state: FileHandleState<Client, Prefetcher>,
+    read_only: bool,
+    write_only: bool,
 }
 
-enum FileHandleType<Client, Prefetcher>
+struct FileHandleState<Client, Prefetcher>
```

Comment:
A few issues with this state:
* Shouldn't the mutex be outside the state struct?
* Is the ETag part of the state? Does it ever change? Shouldn't it change (to None) when an O_RDWR open on an existing file "becomes" a write?
* It would seem natural for this to be an enum (maybe something like: Read/Write/StillUnknown). Is there a reason not to do that? If so, it would help to document it.

Reply:
Agree with the first two, the mutex should cover the state struct and the Etag should be updated.

One reason for implementing it this way is complexity, I have tried the Read/Write/StillUnknown enum approach and it requires us to handle the StillUnknown case in both read and write path. Another reason is that I think we probably want to allow read after write on the same file handle if we open it with O_RDWR flag.

Reply:
I'm still not entirely convinced. The ambiguity of maintaining both fields may only push the complexity down the stack - see for example UploadState::NotStarted - and still make top-level methods more brittle (e.g. what if we "forget" to check for read-only in `FileSystem::write`?).

An enum here would instead always force a check in the top-level read/write methods, but I'd argue that is the right place and where we actually want to transition from StillUnknown to Read or Write. Happy to discuss further though!

Reply:
About your last point, I don't see an enum as a blocker for implementing read after write at some point in the future: it would be a transition from Write to Read, right?